File: zlcore/src/main/java/com/zaitunlabs/zlcore/utils/ViewEnablerUtils.java
package com.zaitunlabs.zlcore.utils;

import android.view.View;

import java.lang.ref.WeakReference;

/**
 * Created by ahsai on 7/28/2018.
 */

public class ViewEnablerUtils extends ReachLostUtils {
    private WeakReference<View> targetViewRef;

    public ViewEnablerUtils(View targetView, int targetReportTotal){
        this.targetViewRef = new WeakReference<View>(targetView);
        setTargetedDoneTotal(targetReportTotal);
        setActionWhenReachTarget(new Runnable() {
            @Override
            public void run() {
                if(ViewEnablerUtils.this.targetViewRef != null) {
                    View targetView = ViewEnablerUtils.this.targetViewRef.get();
                    if(targetView != null) {
                        targetView.setEnabled(true);
                    }
                }
            }
        });
        setActionWhenLostTarget(new Runnable() {
            @Override
            public void run() {
                if(ViewEnablerUtils.this.targetViewRef != null) {
                    View targetView = ViewEnablerUtils.this.targetViewRef.get();
                    if(targetView != null) {
                        targetView.setEnabled(false);
                    }
                }
            }
        });
    }

    @Override
    public void init() {
        if(targetedDoneTotal > 0){
            if(ViewEnablerUtils.this.targetViewRef != null) {
                View targetView = ViewEnablerUtils.this.targetViewRef.get();
                if(targetView != null) {
                    targetView.setEnabled(false);
                }
            }
        }
    }
}

File: zlcore/src/main/java/com/zaitunlabs/zlcore/utils/ViewEnablerUtils.kt
package com.zaitunlabs.zlcore.utils

import android.view.View

import java.lang.ref.WeakReference

/**
 * Created by ahsai on 7/28/2018.
 */

class ViewEnablerUtils(targetView: View, targetReportTotal: Int) : ReachLostUtils() {
    private val targetViewRef: WeakReference<View>?

    init {
        this.targetViewRef = WeakReference(targetView)
        setTargetedDoneTotal(targetReportTotal)
        setActionWhenReachTarget {
            if (this@ViewEnablerUtils.targetViewRef != null) {
                val targetView = this@ViewEnablerUtils.targetViewRef.get()
                if (targetView != null) {
                    targetView.isEnabled = true
                }
            }
        }
        setActionWhenLostTarget {
            if (this@ViewEnablerUtils.targetViewRef != null) {
                val targetView = this@ViewEnablerUtils.targetViewRef.get()
                if (targetView != null) {
                    targetView.isEnabled = false
                }
            }
        }
    }

    override fun init() {
        if (targetedDoneTotal > 0) {
            if (this@ViewEnablerUtils.targetViewRef != null) {
                val targetView = this@ViewEnablerUtils.targetViewRef.get()
                if (targetView != null) {
                    targetView.isEnabled = false
                }
            }
        }
    }
}
